docs/src/prototypes/CopyToClipboard/CopyToClipboard.tsx

```diff
@@ -0,0 +1,135 @@
+import * as React from 'react'
+import { Popup, Button, Text, Portal, Ref } from '@stardust-ui/react/src'
+import * as copyToClipboard from 'copy-to-clipboard'
+
+export type CopyToClipboardProps = {
+  value: string
+  copyPrompt: string
+  timeout?: number
+  attached?: boolean
+  pointing?: boolean
+}
+
+class CopyToClipboard extends React.Component<CopyToClipboardProps> {
+  state = {
+    copied: false,
+    copiedHeight: 0,
+    copiedWidth: 0,
+  }
+  timeoutId = undefined
+  copiedTextRef = React.createRef<HTMLElement>()
+
+  static defaultProps = {
+    timeout: 4000,
+    attached: false,
+    pointing: false,
+  }
+
+  handleClick = () => {
+    if (this.timeoutId !== undefined) {
+      clearTimeout(this.timeoutId)
+    }
+    const { timeout, value } = this.props
+    this.setState({
+      copied: true,
+    })
+    this.timeoutId = setTimeout(() => {
+      this.setState({ copied: false })
+    }, timeout)
+    copyToClipboard(value)
+  }
+
+  componentDidMount() {
+    if (this.copiedTextRef.current === null) {
+      return
+    }
+    this.setState({
+      copiedHeight: this.copiedTextRef.current.clientHeight,
+      copiedWidth: this.copiedTextRef.current.clientWidth,
+    })
+  }
+
+  render() {
+    const { attached, pointing, copyPrompt } = this.props
+    const hideContent = this.state.copied && !attached
+    const copiedStr = 'Copied to clipboard'
+    const copiedText = hideContent ? (
+      <Text size="larger" styles={{ padding: '5px 5px 5px 5px' }}>
+        {copiedStr}
+      </Text>
+    ) : (
+      <Text>{copiedStr}</Text>
+    )
+    const copyText = <Text>{copyPrompt}</Text>
+    let popupContent = this.state.copied ? copiedText : copyText
+    if (hideContent) {
+      popupContent = <Text />
+    }
+    const button = (
+      <Button
+        icon="clipboard-copied-to"
+        iconOnly
+        onClick={this.handleClick}
+        variables={siteVariables => ({
+          colorHover: !this.state.copied && siteVariables.colors.brand[400],
+        })}
+      />
+    )
+    const hiddenPopupVariables = {
+      contentBackgroundColor: 'transparent',
+    }
+    const hiddenPopupContentVariables = {
+      boxShadow: 'none',
+      borderColor: 'transparent',
+    }
+    const portalContentStyle = {
+      display: 'flex',
+      alignItems: 'center',
+      justifyContent: 'center',
+      position: 'fixed' as 'fixed',
+      backgroundColor: 'black',
+      color: 'white',
+      zIndex: 1000,
+      top: `calc(50% - ${this.state.copiedHeight / 2}px)`,
+      left: `calc(50% - ${this.state.copiedWidth / 2}px)`,
+    }
```

Comment:
I believe that we can start it in more simple way and we will not need a ref 😎 

I prepared CodeSandbox: https://codesandbox.io/s/busy-pond-9v49o

Reply:
if the button would be inside of a popup or dialog, would it still be centered relative to the window or relative to the popup/dialog? The requirement is to center it relative to the window (if not attached).

Next level: if the button is in a popup and popup would be closed, would the notification popup be closed as well?

Reply:
> if the button would be inside of a popup or dialog, would it still be centered relative to the window or relative to the popup/dialog?

It will be rendered inside `Portal`, so it will be centered in window.

***



> Next level: if the button is in a popup and popup would be closed, would the notification popup be closed as well?

Yep, like a nested `Popup` because React's Tree will be unmounted.

Reply:
so this might end up being a Notificaton provider/consumer, something like (pseudocode)

```jsx
const notifications = useContext(Notifications)
const ref = useRef();

const handleClick = () => {
  randomAction()
    .then(() => notifications.show( { content: 'Copied', attachTo: ref} ))
    .fail(() => notifications.show( { content: 'Failed to copy', attachTo: ref} ))
}

return (
  <Ref innerRef={ref}><Button content='copy to clipboard' onClick={handleClick}/></Ref>
)
```

then the provider would just render a tooltip (using target prop).

Reply:
@layershifter the consequence of setting `right` and `bottom` props is that user can no longer click on any other elements than the centered message